File: src/main/java/com/mikea/bayes/SumProduct.java
package com.mikea.bayes;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Set;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;
import static com.google.common.collect.Lists.newArrayList;
import static com.google.common.collect.Sets.newHashSet;
import static com.mikea.bayes.VarSet.newVarSet;

/**
 * @author mike.aizatsky@gmail.com
 */
public final class SumProduct {
    public static final VarOrderStrategy DEFAULT_STRATEGY = new MinNeighborsStrategy();

    public static Factor sumProductVariableElimination(
            Iterable<Var> vars,
            List<Factor> factors) {
        return sumProductVariableElimination(vars, factors, DEFAULT_STRATEGY);
    }

    public static Factor sumProductVariableElimination(
            Iterable<Var> vars,
            List<Factor> factors,
            VarOrderStrategy strategy) {
        return sumProductVariableElimination(ProbabilitySpace.fromVars(vars), vars, factors, strategy);
    }

    private static Factor sumProductVariableElimination(ProbabilitySpace space, Iterable<Var> vars, List<Factor> factors, VarOrderStrategy strategy) {
        Set<Var> varSet = newHashSet(vars);
        while (!varSet.isEmpty()) {
            System.out.println("varSet.size() = " + varSet.size());
            Var var = strategy.pickVar(space, varSet, Factor.getScopes(factors));
            factors = sumProductEliminateVar(factors, var);
            varSet.remove(var);
        }

        return Factor.product(factors);
    }

    private static List<Factor> sumProductEliminateVar(Iterable<Factor> factors, Var var) {
        List<Factor> result = newArrayList();
        List<Factor> product = newArrayList();
        for (Factor factor : factors) {
            VarSet scope = factor.getScope();
            if (scope.contains(var)) {
                product.add(factor);
            } else {
                result.add(factor);
            }
        }

        Factor newFactor = Factor.product(product).marginalize(newVarSet(var));
        System.out.println("newFactor.getCardinality = " + newFactor.getScope().getCardinality());
        result.add(newFactor);
        return result;
    }

    public interface VarOrderStrategy {
        Var pickVar(ProbabilitySpace space, Set<Var> vars, Iterable<VarSet> factorScopes);
    }

    public abstract static class GreedyOrderStrategy implements VarOrderStrategy {
        public abstract void computeCosts(long[] costs, Set<Var> vars, Iterable<VarSet> factorScopes);

        @Override
        public Var pickVar(ProbabilitySpace space, Set<Var> vars, Iterable<VarSet> factorScopes) {
            long[] costs = new long[space.getNumVars()];
            Arrays.fill(costs, 1);

            computeCosts(costs, vars, factorScopes);

            System.out.println("Arrays.toString(costs) = " + Arrays.toString(costs));

            Var minVar = null;
            long minCost = Long.MAX_VALUE;

            for (Var var : vars) {
                long cost = costs[var.getIndex()];
                if (cost < 0) cost = Long.MAX_VALUE;
                if (cost < minCost) {
                    minCost = cost;
                    minVar = var;
                }
            }
            System.out.println("minCost = " + minCost);
            System.out.println("minVar = " + minVar);

            return checkNotNull(minVar);
        }
    }

    public static class MinNeighborsStrategy extends GreedyOrderStrategy {
        @Override
        public void computeCosts(long[] costs, Set<Var> vars, Iterable<VarSet> factorScopes) {
            Multimap<Var, Var> neighbors = HashMultimap.create();

            for (VarSet scope : factorScopes) {
                for (Var var : scope) {
                    neighbors.putAll(var, scope);
                }
            }

            for (Var var : neighbors.keySet()) {
                costs[var.getIndex()] = neighbors.get(var).size();
            }
        }
    }

    public static class MinWeightStrategy extends GreedyOrderStrategy {
        @Override
        public void computeCosts(long[] costs, Set<Var> vars, Iterable<VarSet> factorScopes) {
            for (VarSet scope : factorScopes) {
                int factorCardinality = scope.getCardinality();
                checkState(factorCardinality != 0, "Zero cardinality factor: %s", scope);

                for (Var var : scope) {
                    costs[var.getIndex()] *= factorCardinality;
                }
            }
        }
    }

    public static class MinFillStrategy extends GreedyOrderStrategy {
        @Override
        public void computeCosts(long[] costs, Set<Var> vars, Iterable<VarSet> factorScopes) {
            Multimap<Var, Var> neighbors = HashMultimap.create();

            for (VarSet scope : factorScopes) {
                for (Var var : scope) {
                    neighbors.putAll(var, scope);
                }
            }

            for (Var var : neighbors.keySet()) {
                Collection<Var> varNeighbors = neighbors.get(var);

                int count = 0;
                for (Var var1 : varNeighbors) {
                    for (Var var2 : varNeighbors) {
                        if (var1.equals(var2) || var1.equals(var) || var2.equals(var)) continue;

                        if (!neighbors.containsEntry(var1, var2)) {
                            count++;
                        }
                    }
                }

                costs[var.getIndex()] = count / 2;
            }
        }
    }
}

File: src/main/java/com/mikea/bayes/SumProduct.scala
package com.mikea.bayes

import scala.collection.mutable


object SumProduct {

  val DEFAULT_STRATEGY = new MinNeighborsStrategy()

  def sumProductVariableElimination(vars: Iterable[Var], factors: Seq[Factor]): Factor = {
    sumProductVariableElimination(vars, factors, DEFAULT_STRATEGY)
  }

  def sumProductVariableElimination(vars: Iterable[Var], factors: Seq[Factor], strategy: VarOrderStrategy): Factor = {
    sumProductVariableElimination(ProbabilitySpace.fromVars(vars), vars, factors, strategy)
  }

  private def sumProductVariableElimination(space: ProbabilitySpace,
                                            vars: Iterable[Var],
                                            factors: Seq[Factor],
                                            strategy: VarOrderStrategy): Factor = {
    var varSet: Set[Var] = vars.toSet
    var currentFactors: Seq[Factor] = factors
    while (!varSet.isEmpty) {
      println("varSet.size() = " + varSet.size)
      val v = strategy.pickVar(space, varSet, Factor.getScopes(currentFactors))
      currentFactors = sumProductEliminateVar(currentFactors, v)
      varSet -= v
    }
    Factor.product(currentFactors)
  }

  private def sumProductEliminateVar(factors: Iterable[Factor], v: Var): List[Factor] = {
    var result = List[Factor]()
    var product = List[Factor]()
    for (factor <- factors) {
      val scope = factor.scope
      if (scope.contains(v)) {
        product :+= factor
      } else {
        result :+= factor
      }
    }
    val newFactor = Factor.product(product).marginalize(VarSet(v))
    result :+= newFactor
    result
  }

  trait VarOrderStrategy {
    def pickVar(space: ProbabilitySpace, vars: Set[Var], factorScopes: Iterable[VarSet]): Var
  }

  abstract class GreedyOrderStrategy extends VarOrderStrategy {
    def computeCosts(costs: Array[Long], vars: Set[Var], factorScopes: Iterable[VarSet])

    override def pickVar(space: ProbabilitySpace, vars: Set[Var], factorScopes: Iterable[VarSet]): Var = {
      val costs = Array.fill(space.numVars)(1l)
      computeCosts(costs, vars, factorScopes)
      var minVar: Var = null
      var minCost = Long.MaxValue
      for (v <- vars) {
        var cost = costs(v.getIndex)
        if (cost < 0) cost = Long.MaxValue
        if (cost < minCost) {
          minCost = cost
          minVar = v
        }
      }
      println("minCost = " + minCost)
      println("minVar = " + minVar)
      assert(minVar != null)
      minVar
    }
  }

  class MinNeighborsStrategy extends GreedyOrderStrategy {
    override def computeCosts(costs: Array[Long], vars: Set[Var], factorScopes: Iterable[VarSet]) {
      val neighbors = new mutable.HashMap[Var, mutable.Set[Var]]() with mutable.MultiMap[Var, Var]

      for (scope <- factorScopes; v1 <- scope; v2 <- scope) {
        neighbors.addBinding(v1, v2)
      }

      for (v <- neighbors.keySet) {
        costs(v.getIndex) = neighbors.get(v).size
      }
    }
  }

  class MinWeightStrategy extends GreedyOrderStrategy {

    override def computeCosts(costs: Array[Long], vars: Set[Var], factorScopes: Iterable[VarSet]) {
      for (scope <- factorScopes) {
        val factorCardinality = scope.cardinality
        assert(factorCardinality != 0, s"Zero cardinality factor: $scope")
        for (v <- scope) {
          costs(v.getIndex) *= factorCardinality
        }
      }
    }
  }

  class MinFillStrategy extends GreedyOrderStrategy {

    override def computeCosts(costs: Array[Long], vars: Set[Var], factorScopes: Iterable[VarSet]) {
      ???
/*
      val neighbors = HashMultimap.create()
      for (scope <- factorScopes; `var` <- scope) {
        neighbors.putAll(`var`, scope)
      }
      for (`var` <- neighbors.keySet) {
        val varNeighbors = neighbors.get(`var`)
        var count = 0
        for (var1 <- varNeighbors; var2 <- varNeighbors) {
          if (var1 == var2 || var1 == `var` || var2 == `var`) //continue
            if (!neighbors.containsEntry(var1, var2)) {
              count += 1
            }
        }
        costs(`var`.getIndex) = count / 2
      }
*/
    }
  }
}
